# EnsembleSeq
In-silico components of the EnsembleSeq workflow

## How to run?
### 1. Raredynamics component of EnsembleSeq i.e. for real time monitoring of species saturation
#### Assuming the scripts/codes of this repo are in the same folder as the fastq files generated by the sequencing machine for a given barcode, only the following needs to be run:

    sh raredynamics.sh

> Access the code here: [raredynamics.sh](https://github.com/sunilnagpal/EnsembleSeq/blob/main/raredynamics.sh)

> *Following is the schematic representation of the dynamic monitoring of species saturation that can guide the optimal run-time for nanopore sequencing of ensembled amplicons. T1,T2,T3..Tn represent various time points after starting the sequencing run, where T1 refers to the time when first set of sequences are real-time base called and written on the disc. Subsequent time points refer to the events of new sequences being written on disc (for e.g. nanopore guppy basecaller is configured to write 4000 reads in each writing cycle by default). For each time point, all available reads can be concatenated to a single fastq file and the reads can be subjected to direct taxonomic classification (fast but approximate) using RDP classifier, employing both bacterial and fungal databases. The accumulated classified sequences can then be separately analysed for species saturation through rarefaction analysis (using iNEXT library of R programming language) and the rarefaction plot for the given time point can be plotted using ggplot2 library of R.  Taxonomic classifications of independent time point specific sequences (without sequence accumulation) can also be aggregated to the accumulated taxonomic composition of the previous time point to yied the similar temporal trend of rarefaction.* 

<img src="https://github.com/sunilnagpal/EnsembleSeq/blob/main/img/raredynamics.png" width=50% height=50%>



