# EnsembleSeq
This repository provides access to the in-silico components of the EnsembleSeq workflow. For details pertaining to the in-vitro components, please refer the manuscript. Briefly, following schematic gives an overview of the overall workflow of EnsembleSeq:

<img src="https://github.com/sunilnagpal/EnsembleSeq/blob/main/img/EnsembleSeq_workflow.jpg" width=60% height=60%>


## How to run the in-silico component?
### 1. Raredynamics component of EnsembleSeq i.e. for real time monitoring of species saturation
#### Assuming the scripts/codes of this repo are in the same folder as the fastq files generated by the sequencing machine for a given barcode, only the following needs to be run:

    sh raredynamics.sh

> Access the code here: [raredynamics.sh](https://github.com/sunilnagpal/EnsembleSeq/blob/main/raredynamics.sh)

> *Following is the schematic representation of the dynamic monitoring of species saturation that can guide the optimal run-time for nanopore sequencing of ensembled amplicons. T1,T2,T3..Tn represent various time points after starting the sequencing run, where T1 refers to the time when first set of sequences are real-time base called and written on the disc. Subsequent time points refer to the events of new sequences being written on disc (for e.g. nanopore guppy basecaller is configured to write 4000 reads in each writing cycle by default). For each time point, all available reads can be concatenated to a single fastq file and the reads can be subjected to direct taxonomic classification (fast but approximate) using RDP classifier, employing both bacterial and fungal databases. The accumulated classified sequences can then be separately analysed for species saturation through rarefaction analysis (using iNEXT library of R programming language) and the rarefaction plot for the given time point can be plotted using ggplot2 library of R.  Taxonomic classifications of independent time point specific sequences (without sequence accumulation) can also be aggregated to the accumulated taxonomic composition of the previous time point to yied the similar temporal trend of rarefaction.* 

<img src="https://github.com/sunilnagpal/EnsembleSeq/blob/main/img/raredynamics.png" width=50% height=50%>

### 2. Post completion of run, High accuracy (HAC) (re)basecalling
#### Once the sequencing has completed, a high accuracy basecalling is recommended (depending on the GPU availability, super high accuracy (sup) models may also be used. Following command ensures that barcodes are trimmed, adapters are trimmed, chimeras are avoided:
    guppy_basecaller.exe -i <PATH_raw> -s <PATH_store> -c dna_r10.4.1_e8.2_400bps_5khz_hac.cfg -x 'auto' --recursive --do_read_splitting --barcode_kits SQK-NBD114-96 --detect_barcodes --enable_trim_barcodes
Here, 
> **PATH_raw** represents the path to raw POD5 or FAST5 files

> **PATH_store** represents the path where rebasecalled sequences should be stored

> **dna_r10.4.1_e8.2_400bps_5khz_hac.cfg** is the model to be chosen for rebasecalling. Replace it as per your flow cell version and accuracy level (fast, hac, sup being the three options)

### 3. Primer based demultiplexing of kingdoms
#### Once the sequences have been basecalled, separate reads into bacteriome and mycobiome (or other kingdoms) based on the primers used for amplicon generation. This step requires [cutadapt](https://cutadapt.readthedocs.io/en/stable/).
E.g. for Mycobiome:
    
    cutadapt -g XTCCGTAGGTGAACCTGCGG  -a TCCGTAGGTGAACCTGCGGX -a TCCTCCGCTTATTGATATGCX -g XTCCTCCGCTTATTGATATGC -o cutted_fungi_barcode1.fastq barcode1.fastq --untrimmed-output untrimmed_fungi_barcode1.fq --revcomp -O 15

E.g. for Bacteriome:

    cutadapt -g XAGRGTTYGATYMTGGCTCAG  -a AGRGTTYGATYMTGGCTCAGX -a CGGYTACCTTGTTACGACTTX -g XCGGYTACCTTGTTACGACTT -o cutted_bacteria_barcode1.fastq barcode1.fastq --untrimmed-output untrimmed_bacteria_barcode1.fq --revcomp -O 15

This can be repeated for all barcode specific reads.
>Note: During rebasecalling, multiple fastq files are created in each barcode specific directory. Ensure to concatenate them all into a single barcode specific fastq file. You can now assess quality of the sequences specific to bacteriome and mycobiome. Trim the reads to appropriate length or filter reads by their length and quality using: 

### 4. Species level taxonomic classification
#### Once the kingdom specific sequences have been segegated into corresponding directories, next step is to perform taxonomic classification. This step requires [EMU](https://gitlab.com/treangenlab/emu).
E.g. for Mycobiome:

    ./emu abundance --keep-counts --keep-read-assignments inputsfungilengthtrimmed/Barcode.1.fasta --output-dir ./fungalresultslengthtrimmed_milliondb --db  unitemilliondb
    
E.g. for Bacteriome:

    ./emu abundance --keep-counts --keep-read-assignments inputsbacterialengthtrimmed/Barcode.1.fasta --output-dir ./bacteriaresultslengthtrimmed_16s --db  emu_database

Here **unitemilliondb** is the database we have created using **6499364** ITS sequences in the Full UNITE+INSD dataset for Fungi. Version 18.07.2023.

**emu_database** is by-default available for 16S taxonomic classification in the latest release of EMU
